<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1">

<meta name="author" content="Lai Jiang" />

<meta name="date" content="2020-01-29" />

<title>hJAM Vignette</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<link href="data:text/css;charset=utf-8,body%20%7B%0Abackground%2Dcolor%3A%20%23fff%3B%0Amargin%3A%201em%20auto%3B%0Amax%2Dwidth%3A%20700px%3B%0Aoverflow%3A%20visible%3B%0Apadding%2Dleft%3A%202em%3B%0Apadding%2Dright%3A%202em%3B%0Afont%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Afont%2Dsize%3A%2014px%3B%0Aline%2Dheight%3A%201%2E35%3B%0A%7D%0A%23header%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0A%23TOC%20%7B%0Aclear%3A%20both%3B%0Amargin%3A%200%200%2010px%2010px%3B%0Apadding%3A%204px%3B%0Awidth%3A%20400px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Aborder%2Dradius%3A%205px%3B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Afont%2Dsize%3A%2013px%3B%0Aline%2Dheight%3A%201%2E3%3B%0A%7D%0A%23TOC%20%2Etoctitle%20%7B%0Afont%2Dweight%3A%20bold%3B%0Afont%2Dsize%3A%2015px%3B%0Amargin%2Dleft%3A%205px%3B%0A%7D%0A%23TOC%20ul%20%7B%0Apadding%2Dleft%3A%2040px%3B%0Amargin%2Dleft%3A%20%2D1%2E5em%3B%0Amargin%2Dtop%3A%205px%3B%0Amargin%2Dbottom%3A%205px%3B%0A%7D%0A%23TOC%20ul%20ul%20%7B%0Amargin%2Dleft%3A%20%2D2em%3B%0A%7D%0A%23TOC%20li%20%7B%0Aline%2Dheight%3A%2016px%3B%0A%7D%0Atable%20%7B%0Amargin%3A%201em%20auto%3B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dcolor%3A%20%23DDDDDD%3B%0Aborder%2Dstyle%3A%20outset%3B%0Aborder%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0Aborder%2Dwidth%3A%202px%3B%0Apadding%3A%205px%3B%0Aborder%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dstyle%3A%20inset%3B%0Aline%2Dheight%3A%2018px%3B%0Apadding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0Aborder%2Dleft%2Dstyle%3A%20none%3B%0Aborder%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Ap%20%7B%0Amargin%3A%200%2E5em%200%3B%0A%7D%0Ablockquote%20%7B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Apadding%3A%200%2E25em%200%2E75em%3B%0A%7D%0Ahr%20%7B%0Aborder%2Dstyle%3A%20solid%3B%0Aborder%3A%20none%3B%0Aborder%2Dtop%3A%201px%20solid%20%23777%3B%0Amargin%3A%2028px%200%3B%0A%7D%0Adl%20%7B%0Amargin%2Dleft%3A%200%3B%0A%7D%0Adl%20dd%20%7B%0Amargin%2Dbottom%3A%2013px%3B%0Amargin%2Dleft%3A%2013px%3B%0A%7D%0Adl%20dt%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Aul%20%7B%0Amargin%2Dtop%3A%200%3B%0A%7D%0Aul%20li%20%7B%0Alist%2Dstyle%3A%20circle%20outside%3B%0A%7D%0Aul%20ul%20%7B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Apre%2C%20code%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0Aborder%2Dradius%3A%203px%3B%0Acolor%3A%20%23333%3B%0Awhite%2Dspace%3A%20pre%2Dwrap%3B%20%0A%7D%0Apre%20%7B%0Aborder%2Dradius%3A%203px%3B%0Amargin%3A%205px%200px%2010px%200px%3B%0Apadding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Acode%20%7B%0Afont%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0Afont%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0Apadding%3A%202px%200px%3B%0A%7D%0Adiv%2Efigure%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0Abackground%2Dcolor%3A%20%23FFFFFF%3B%0Apadding%3A%202px%3B%0Aborder%3A%201px%20solid%20%23DDDDDD%3B%0Aborder%2Dradius%3A%203px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Amargin%3A%200%205px%3B%0A%7D%0Ah1%20%7B%0Amargin%2Dtop%3A%200%3B%0Afont%2Dsize%3A%2035px%3B%0Aline%2Dheight%3A%2040px%3B%0A%7D%0Ah2%20%7B%0Aborder%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Apadding%2Dbottom%3A%202px%3B%0Afont%2Dsize%3A%20145%25%3B%0A%7D%0Ah3%20%7B%0Aborder%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Afont%2Dsize%3A%20120%25%3B%0A%7D%0Ah4%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0Amargin%2Dleft%3A%208px%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Ah5%2C%20h6%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23ccc%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Aa%20%7B%0Acolor%3A%20%230033dd%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Aa%3Ahover%20%7B%0Acolor%3A%20%236666ff%3B%20%7D%0Aa%3Avisited%20%7B%0Acolor%3A%20%23800080%3B%20%7D%0Aa%3Avisited%3Ahover%20%7B%0Acolor%3A%20%23BB00BB%3B%20%7D%0Aa%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0Aa%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%20code%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%0A" rel="stylesheet" type="text/css" />




</head>

<body>




<h1 class="title toc-ignore">hJAM Vignette</h1>
<h4 class="author">Lai Jiang</h4>
<h4 class="date">2020-01-29</h4>



<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(hJAM)</code></pre></div>
<div id="overview" class="section level2">
<h2>Overview</h2>
<p><em>hJAM</em> is a package developed to implement the hJAM model, which is designed to estimate the associations between multiple intermediates and outcome in a Mendelian Randomization or Transcriptome analysis.</p>
<p>Mendelian randomization (MR) and transcriptome-wide association studies (TWAS) can be viewed as the same approach within the instrumental variable analysis framework using genetic variants. They differ in their intermediates: MR focuses on modifible risk factors while TWAS focuses on gene expressions. We can use a two-stage hierarchical model to unify the framework of MR and TWAS. Details are described in our paper.</p>
</div>
<div id="implementation" class="section level2">
<h2>Implementation</h2>
<p>We have two methods in our pacakge:</p>
<ul>
<li><em>hJAM</em> with linear regression: <code>hJAM_lnreg</code></li>
<li><em>hJAM</em> with Egger regression: <code>hJAM_egger</code></li>
</ul>
<div id="general-input" class="section level4">
<h4>General input</h4>
<p>The input of the two <em>hJAM</em> model includes:</p>
<ul>
<li><code>betas.Gy</code> is the beta vector of the marginal effects of SNPs on the outcome. It can be directly extracted from a GWAS where the outcome is the outcome of interests.</li>
<li><code>N.Gy</code> is the sample size of the GWAS where the users extract the <code>betas.Gy</code>.</li>
<li><code>Gl</code> is the reference genotype matrix with number of columns equals to the number of SNPs in the instrument set. It can be a publicly available genotype data, such as <a href="https://www.internationalgenome.org/data/">1000 Genomes project</a>. Users need to confirm the genotype matrix is in a dosage format. For VCF files, users could use <a href="http://vcftools.sourceforge.net/">vcftool</a> to convert it into dosage format.</li>
<li><code>A</code> is the alpha matrix of the conditional effects of SNPs on the intermediates. The number of columns of <code>A</code> equals the number of intermediates in the usersâ€™ research question. To conditional <code>A</code> matrix can be converted from a marginal <code>A</code> by using <code>get_cond_A</code> function in hJAM package. We will describe this later.</li>
<li><code>ridgeTerm</code> is the ridge term that we add to the diagonal of Cholesky decomposition component matrix <span class="math inline">\(L\)</span> to enforce it to be a positive definite matrix. Please see details in our paper.</li>
</ul>
</div>
<div id="get-conditional-hata-matrix" class="section level4">
<h4>Get conditional <span class="math inline">\(\hat{A}\)</span> matrix</h4>
<p>To generate a conditional estimate <span class="math inline">\(\hat{A}\)</span> matrix from a marginal estimate <span class="math inline">\(\hat{A}\)</span> matrix, users can use the <code>get_cond_A</code> (if number of intermediates &gt; 1) or <code>get_cond_alpha</code> (if number of intermediate = 1) functions in hJAM package. Examples are given in next section.</p>
<p>For MR questions, the intermediates are modifiable risk factors. The marginal <span class="math inline">\(\hat{A}\)</span> can be extracted from different GWAS whose the outcomes are the risk factors of interests. For example, for intermediate as body mass index, the marginal <span class="math inline">\(\hat{\alpha}\)</span> vector can be extracted from the <a href="https://portals.broadinstitute.org/collaboration/giant/index.php/GIANT_consortium_data_files">GIANT consortium</a>.</p>
<p>For TWAS questions, the intermediates are gene expressions. There are two ways to obtain the elements in <span class="math inline">\(\hat{A}\)</span> matrix.</p>
<ol style="list-style-type: decimal">
<li><p><a href="https://gtexportal.org/home/">GTEx portal</a>: the GTEx project provides marginal summary statistics between SNPs and gene expressions in different tissues.</p></li>
<li><p><a href="http://predictdb.org/">PredictDB</a>: the PredictDB is developed by the PrediXcan group. It uses elastic net on individual level data from the GTEx project.</p></li>
</ol>
<hr />
<p><strong>Implementation with caution:</strong></p>
<ul>
<li>The users need to confirm that the reference allele of each SNP is the same for all estimates in <span class="math inline">\(\hat{A}\)</span> matrix and <span class="math inline">\(\hat{\beta}\)</span> vector.</li>
<li>If the users want to run the <code>hJAM_egger</code> function, make sure that the directions of the association estimates in <span class="math inline">\(\hat{A}\)</span> matrix are positive. It is possible that there are some of SNPs cannot be positive due to the reverse effects between intermediates on the outcome.</li>
</ul>
<hr />
</div>
</div>
<div id="example" class="section level2">
<h2>Example</h2>
<p>In our package, we prepared a data example which we have described in detail in our paper. In this data example, we focus on the conditional effects of body mass index (BMI) and type 2 diabetes (T2D) on myocardial infarction (MI).</p>
<p>We identified 75 and 136 significantly BMI- and T2D-associated SNPs from GIANT consortium and DIAGRAM+GERA+UKB, respectively. In this set of SNPs, there was one overlapping SNP in both the instrument sets for BMI and T2D. In total, we have 210 SNPs identified. The association estimates between the 210 SNPs and MI were collected from UK Biobank.</p>
<div id="data-exploration" class="section level4">
<h4>Data exploration</h4>
<p>A quick look at the data in the example -</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">data</span>(<span class="st">&quot;conditional_A&quot;</span>)
<span class="kw">data</span>(<span class="st">&quot;betas.Gy&quot;</span>)
<span class="kw">data</span>(<span class="st">&quot;SNPs_info&quot;</span>)
conditional_A[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, ]</code></pre></div>
<pre><code>##                bmi          t2d
##  [1,]  0.019531085  0.072587211
##  [2,]  0.025262061  0.013586392
##  [3,] -0.005147363  0.089673178
##  [4,]  0.046302578  0.041313103
##  [5,]  0.016849395 -0.004564683
##  [6,]  0.012345062  0.036207829
##  [7,]  0.042065546  0.013786703
##  [8,]  0.073772221  0.118575569
##  [9,] -0.004559686  0.051510367
## [10,] -0.012843666  0.091788493</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">betas.Gy[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>]</code></pre></div>
<pre><code>##  [1]  0.0197298348  0.0133151413  0.0008717583  0.0213550014 -0.0031514278
##  [6] -0.0062721407  0.0108936262 -0.0109635708  0.0003266145  0.0077205399</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">SNPs_info[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, ]</code></pre></div>
<pre><code>##           SNP Major_A    ref_frq
## 1   rs2296173       G 0.12079449
## 2    rs657452       A 0.54742602
## 3  rs12088739       A 0.87859749
## 4   rs3101336       C 0.67815160
## 5  rs12566985       G 0.67936765
## 6  rs12401738       A 0.15707337
## 7  rs11165643       T 0.50081070
## 8  rs17024393       C 0.03080665
## 9   rs1127655       C 0.42298338
## 10  rs2493394       G 0.19092015</code></pre>
<p>In this package, we embed two fucntions for the users to check the SNPs they use in the analysis visually:</p>
<ul>
<li>Scatter plot: <code>SNPs_scatter_plot</code></li>
<li>Heatmap: <code>SNPs_heatmap</code></li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">scatter_plot_p =<span class="st"> </span><span class="kw">SNPs_scatter_plot</span>(<span class="dt">A =</span> conditional_A, <span class="dt">betas.Gy =</span> betas.Gy, <span class="dt">num_X =</span> <span class="dv">2</span>)
heatmap_p =<span class="st"> </span><span class="kw">SNPs_heatmap</span>(Gl)</code></pre></div>
</div>
<div id="conversion-of-hata-matrix" class="section level4">
<h4>Conversion of <span class="math inline">\(\hat{A}\)</span> matrix</h4>
<p>You could use function <code>get_cond_A</code> function to run JAM on the marginal estimates <span class="math inline">\(\hat{A}\)</span> matrix and convert it into a conditional estimates <span class="math inline">\(\hat{A}\)</span> matrix.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">data</span>(<span class="st">&quot;marginal_A&quot;</span>)
cond_A =<span class="st"> </span><span class="kw">get_cond_A</span>(<span class="dt">marginal_A =</span> marginal_A, <span class="dt">Gl =</span> Gl, <span class="dt">N.Gx =</span> <span class="dv">339224</span>, <span class="dt">ridgeTerm =</span> T)
cond_A[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, ]</code></pre></div>
<pre><code>##                bmi          t2d
##  [1,]  0.019531085  0.072587211
##  [2,]  0.025262061  0.013586392
##  [3,] -0.005147363  0.089673178
##  [4,]  0.046302578  0.041313103
##  [5,]  0.016849395 -0.004564683
##  [6,]  0.012345062  0.036207829
##  [7,]  0.042065546  0.013786703
##  [8,]  0.073772221  0.118575569
##  [9,] -0.004559686  0.051510367
## [10,] -0.012843666  0.091788493</code></pre>
</div>
<div id="hjam" class="section level4">
<h4>hJAM</h4>
<p>The default version of <em>hJAM</em> restricts the intercept to be zero.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">hJAM<span class="op">::</span><span class="kw">hJAM_lnreg</span>(<span class="dt">betas.Gy =</span> betas.Gy, <span class="dt">N.Gy =</span> <span class="dv">459324</span>, <span class="dt">A =</span> conditional_A, <span class="dt">Gl =</span> Gl, <span class="dt">ridgeTerm =</span> <span class="ot">TRUE</span>) <span class="co"># 459324 is the sample size of the UK Biobank GWAS of MI</span></code></pre></div>
<pre><code>## ------------------------------------------------------ 
##                    hJAM output                         
## ------------------------------------------------------ 
## Number of SNPs used in model: 210 
## 
##     Estimate StdErr         95% CI       Pvalue
## bmi    0.322  0.061 (0.202, 0.442) 3.189944e-07
## t2d    0.119  0.017 (0.085, 0.153) 4.124526e-11
## ------------------------------------------------------</code></pre>
<p>Another method in this package is <em>hJAM</em> with Egger regression, which is analogus to MR egger. It allows the intercept to be non-zero.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">hJAM<span class="op">::</span><span class="kw">hJAM_egger</span>(<span class="dt">betas.Gy =</span> betas.Gy, <span class="dt">N.Gy =</span> <span class="dv">459324</span>, <span class="dt">A =</span> conditional_A, <span class="dt">Gl =</span> Gl, <span class="dt">ridgeTerm =</span> <span class="ot">TRUE</span>) <span class="co"># 459324 is the sample size of the UK Biobank GWAS of MI</span></code></pre></div>
<pre><code>## ------------------------------------------------------ 
##                    hJAM egger output                   
## ------------------------------------------------------ 
## Number of SNPs used in model: 210 
## 
##     Estimate StdErr        95% CI       Pvalue
## bmi    0.302  0.070 (0.163, 0.44) 2.817908e-05
## t2d    0.107  0.027 (0.055, 0.16) 8.175202e-05
## 
## Intercept
##      Est.Int StdErr.Int 95% CI.Int        Pvalue.Int
## [1,] &quot;0.453&quot; &quot;0.787&quot;    &quot;(-1.099, 2.005)&quot; &quot;0.565&quot;   
## ------------------------------------------------------</code></pre>
</div>
</div>
<div id="conclusion" class="section level2">
<h2>Conclusion</h2>
<p>We presented the main usage of <code>hJAM</code> package. For more details about each function, please go check the package documentation. If you would like to give us feedback or report issue, please tell us on <a href="https://github.com/lailyaljiang/hJAM">Github</a>.</p>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
